File: spiral-matrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []

        minrows, mincols = 0, 0
        maxrows, maxcols = len(matrix), len(matrix[0])
        count = maxrows * maxcols
        r, c = 0, -1

        dirs = (0, 1), (1, 0), (0, -1), (-1, 0)
        d = 0

        while count > 0:
            nr, nc = r + dirs[d][0], c + dirs[d][1]

            if nr in range(minrows, maxrows) and nc in range(mincols, maxcols):
                ans.append(matrix[nr][nc])
                r, c = nr, nc
                count -= 1

            else:

                if d == 0:
                    minrows += 1
                elif d == 1:
                    maxcols -= 1
                elif d == 2:
                    maxrows -= 1
                elif d == 3:
                    mincols += 1
                
                d = (d+1)%4

        return ans

File: spiral-matrix.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
		vector<int> ans;

		int minrows = 0;
		int mincols = 0;
		int maxrows = matrix.size();
		int maxcols = matrix[0].size();
		int count = maxrows * maxcols;

		int dirs[][2] = {
			{0, 1}, {1, 0}, {0, -1}, {-1, 0},
		};
		int d = 0;
		int r = 0;
		int c = -1;

		while (count > 0) {
			int nr = r + dirs[d][0];
			int nc = c + dirs[d][1];

			if ((minrows <= nr) && (nr < maxrows) && (mincols <= nc) && (nc < maxcols)) {
				ans.push_back(matrix[nr][nc]);
				r = nr;
				c = nc;
				count -= 1;
			} else {
				if (d == 0) {
					minrows += 1;
				} else if (d == 1) {
					maxcols -= 1;
				} else if (d == 2) {
					maxrows -= 1;
				} else if (d == 3) {
					mincols += 1;
				}

				d = (d+1)%4;
			}
		}

		return ans;
    }
};
